### ISSUE TITLE
Wal-g fails to upload wal to s3 in archive command - works when ran manually

### ISSUE THREAD

**Issue by vfrans** [zalando/postgres-operator#1233]
- **Which image of the operator are you using?** registry.opensource.zalan.do/acid/postgres-operator:v1.5.0
- **Where do you run it - cloud or metal? Kubernetes or OpenShift?** Baremetal
- **Are you running Postgres Operator in production?** no
- **Type of issue?** Bug report

Hello,

I'm trying to configure wal archiving to s3 using postgres-operator. I couldn't really find any documentation about it, so here is what i have tried:

My configuration configmap is like this:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: postgres-operator
data:
  # additional_secret_mount: "some-secret-name"
  # additional_secret_mount_path: "/some/dir"
  api_port: "8080"
  aws_region: eu-west-1
  cluster_domain: cluster.local
  cluster_history_entries: "1000"
  cluster_labels: application:spilo
  cluster_name_label: cluster-name
  # connection_pooler_default_cpu_limit: "1"
  # connection_pooler_default_cpu_request: "500m"
  # connection_pooler_default_memory_limit: 100Mi
  # connection_pooler_default_memory_request: 100Mi
  connection_pooler_image: "registry.opensource.zalan.do/acid/pgbouncer:master-11"
  # connection_pooler_max_db_connections: 60
  # connection_pooler_mode: "transaction"
  # connection_pooler_number_of_instances: 2
  # connection_pooler_schema: "pooler"
  # connection_pooler_user: "pooler"
  # custom_service_annotations: "keyx:valuez,keya:valuea"
  # custom_pod_annotations: "keya:valuea,keyb:valueb"
  db_hosted_zone: db.example.com
  debug_logging: "true"
  # default_cpu_limit: "1"
  # default_cpu_request: 100m
  # default_memory_limit: 500Mi
  # default_memory_request: 100Mi
  # delete_annotation_date_key: delete-date
  # delete_annotation_name_key: delete-clustername
  docker_image: registry.opensource.zalan.do/acid/spilo-12:1.6-p5
  # downscaler_annotations: "deployment-time,downscaler/*"
  # enable_admin_role_for_users: "true"
  # enable_crd_validation: "true"
  # enable_database_access: "true"
  # enable_init_containers: "true"
  # enable_lazy_spilo_upgrade: "false"
  enable_master_load_balancer: "false"
  # enable_pod_antiaffinity: "false"
  # enable_pod_disruption_budget: "true"
  # enable_postgres_team_crd: "false"
  # enable_postgres_team_crd_superusers: "false"
  enable_replica_load_balancer: "false"
  # enable_shm_volume: "true"
  # enable_sidecars: "true"
  # enable_team_superuser: "false"
  enable_teams_api: "false"
  # etcd_host: ""
  external_traffic_policy: "Cluster"
  # gcp_credentials: ""
  # kubernetes_use_configmaps: "false"
  # infrastructure_roles_secret_name: "postgresql-infrastructure-roles"
  # infrastructure_roles_secrets: "secretname:monitoring-roles,userkey:user,passwordkey:password,rolekey:inrole"
  # inherited_labels: application,environment
  # kube_iam_role: ""
  # log_s3_bucket: ""
  logical_backup_docker_image: "registry.opensource.zalan.do/acid/logical-backup"
  # logical_backup_s3_access_key_id: ""
  logical_backup_s3_bucket: "my-bucket-url"
  # logical_backup_s3_region: ""
  # logical_backup_s3_endpoint: ""
  # logical_backup_s3_secret_access_key: ""
  logical_backup_s3_sse: "AES256"
  logical_backup_schedule: "30 00 * * *"
  master_dns_name_format: "{cluster}.{team}.{hostedzone}"
  # master_pod_move_timeout: 20m
  # max_instances: "-1"
  # min_instances: "-1"
  # min_cpu_limit: 250m
  # min_memory_limit: 250Mi
  # node_readiness_label: ""
  # oauth_token_secret_name: postgresql-operator
  # pam_configuration: |
  #  https://info.example.com/oauth2/tokeninfo?access_token= uid realm=/employees
  # pam_role_name: zalandos
  pdb_name_format: "postgres-{cluster}-pdb"
  # pod_antiaffinity_topology_key: "kubernetes.io/hostname"
  pod_deletion_wait_timeout: 10m
  pod_environment_configmap: "default/pod-var-custom"
  # pod_environment_secret: "my-custom-secret"
  pod_label_wait_timeout: 10m
  pod_management_policy: "ordered_ready"
  # pod_priority_class_name: "postgres-pod-priority"
  pod_role_label: spilo-role
  # pod_service_account_definition: ""
  pod_service_account_name: "postgres-pod"
  # pod_service_account_role_binding_definition: ""
  pod_terminate_grace_period: 5m
  # postgres_superuser_teams: "postgres_superusers"
  # protected_role_names: "admin"
  ready_wait_interval: 3s
  ready_wait_timeout: 30s
  repair_period: 5m
  replica_dns_name_format: "{cluster}-repl.{team}.{hostedzone}"
  replication_username: standby
  resource_check_interval: 3s
  resource_check_timeout: 10m
  resync_period: 30m
  ring_log_lines: "100"
  secret_name_template: "{username}.{cluster}.credentials"
  # sidecar_docker_images: ""
  # set_memory_request_to_limit: "false"
  # spilo_runasuser: 101
  # spilo_runasgroup: 103
  # spilo_fsgroup: 103
  spilo_privileged: "false"
  storage_resize_mode: "pvc"
  super_username: postgres
  # team_admin_role: "admin"
  # team_api_role_configuration: "log_statement:all"
  # teams_api_url: http://fake-teams-api.default.svc.cluster.local
  # toleration: ""
  # wal_gs_bucket: ""
  # wal_s3_bucket: ""
  watched_namespace: "*"  # listen to all namespaces
  workers: "8"
```
(Notice that I specify pod_environment_configmap: "default/pod-var-custom")

The configmap default/pod-var-custom is like this:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: pod-var-custom
data:
  BACKUP_SCHEDULE: "0 */12 * * *"
  USE_WALG_BACKUP: "true"
  USE_WALG_RESTORE: "true"
  BACKUP_NUM_TO_RETAIN: "14"
  AWS_ACCESS_KEY_ID: "****"
  AWS_SECRET_ACCESS_KEY: "****"
  AWS_ENDPOINT: "s3.eu-west-1.amazonaws.com"
  AWS_REGION: "eu-west-1"
  WAL_S3_BUCKET: "somebucket"
  WALG_DOWNLOAD_CONCURRENCY: "1"
```

When troubleshooting here is the status of the generated pods:
Wal-g does not seem configured correctly:
```
root@team-cluster-0:/home/postgres# wal-g backup-list
ERROR: 2020/11/25 14:16:25.147566 No storage is configured now, please set one of following settings: [WALG_S3_PREFIX WALG_FILE_PREFIX WALG_GS_PREFIX WALG_AZ_PREFIX WALG_SWIFT_PREFIX]
root@team-cluster-0:/home/postgres# 
```
Here are the env var related to aws|bakup|wal:
```
root@team-cluster-0:/home/postgres# env |grep -i -e 'backup\|aws\|wal'
SPILO_CONFIGURATION={"postgresql":{"bin_dir":"/usr/lib/postgresql/11/bin","parameters":{"checkpoint_timeout":"5min","log_checkpoints":"on","log_connections":"on","log_disconnections":"on","log_min_duration_statement":"1000","log_statement":"ddl","shared_buffers":"256MB"},"pg_hba":["host all all all md5"]},"bootstrap":{"initdb":[{"auth-host":"md5"},{"auth-local":"trust"}],"users":{"zalandos":{"password":"","options":["CREATEDB","NOLOGIN"]}},"dcs":{"postgresql":{"parameters":{"max_connections":"500","wal_log_hints":"on"}}}}}
AWS_SECRET_ACCESS_KEY=****
WAL_S3_BUCKET=somebucket
WALG_VERSION=v0.2.15
BACKUP_SCHEDULE=0 */12 * * *
AWS_REGION=eu-west-1
WALG_DOWNLOAD_CONCURRENCY=1
WALE_VERSION=1.1.0
BACKUP_NUM_TO_RETAIN=14
AWS_ENDPOINT=s3.eu-west-1.amazonaws.com
USE_WALG_RESTORE=true
AWS_ACCESS_KEY_ID=****
WALE_ENV_DIR=/run/etc/wal-e.d/env
USE_WALG_BACKUP=true

```

archive_command is set to:
```
archive_command = 'envdir "/run/etc/wal-e.d/env" wal-g wal-push "%p"'
```

When postgres runs the command, here are the logs in the database:
```
 archive command failed with exit code 1  | The failed archive command was: envdir "/run/etc/wal-e.d/env" wal-g wal-push "pg_wal/000000010000000000000001"
```

However, when I run the same command manually in the container:
```
root@team-cluster-0:/home/postgres/pgdata/pgroot/data#  envdir "/run/etc/wal-e.d/env" wal-g wal-push "pg_wal/000000010000000000000001"
INFO: 2020/11/25 14:51:21.527886 FILE PATH: 000000010000000000000002.000000D0.backup.lz4
INFO: 2020/11/25 14:51:21.622288 FILE PATH: 000000010000000000000001.lz4
root@team-cluster-0:/home/postgres/pgdata/pgroot/data# 
```

This is very frustating as I don't really have a way to change the archive_command to get the output of wal-g.

Would have any idea?

Thanks a lot in advance,

Best regards,

**Comment by vfrans** [comment]
I tried to execute the command:
```
envdir "/run/etc/wal-e.d/env" wal-g wal-push "pg_wal/000000010000000000000001"
```
but this time, as postgres, not as root as in my previous debug, i'm getting closer:
```
postgres@team-cluster-0:~/pgdata/pgroot/data$ envdir "/run/etc/wal-e.d/env" wal-g wal-push "pg_wal/000000010000000000000001"
ERROR: 2020/11/26 04:10:00.193819 failed to configure folder: failed to create new session: AWS_REGION is not set and s3:GetBucketLocation failed: AuthorizationHeaderMalformed: The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'eu-west-1'
        status code: 400, request id: 0BC25F2FAFDD6494, host id: Ai82u584hLaeW2ko1m3UJiL5hsAkNnGG+F2GH1FKIriFbqJBJ0sQSuXB01Vd6Wjw7Dj0LYXicGs=
postgres@team-cluster-0:~/pgdata/pgroot/data$ 
```

Indeed, AWS_REGION is not present in the wal-e env dir:
```
postgres@team-cluster-0:~/pgdata/pgroot/data$ ls /run/etc/wal-e.d/env
AWS_ACCESS_KEY_ID  AWS_ENDPOINT  AWS_SECRET_ACCESS_KEY  BACKUP_NUM_TO_RETAIN  TMPDIR  USE_WALG_BACKUP  USE_WALG_RESTORE  WALE_LOG_DESTINATION  WALE_S3_ENDPOINT  WALE_S3_PREFIX  WALG_DISABLE_S3_SSE  WALG_DOWNLOAD_CONCURRENCY  WALG_S3_PREFIX  WALG_UPLOAD_CONCURRENCY
postgres@team-cluster-0:~/pgdata/pgroot/data$ 
```

I can't really locate the part of the code that does this but it seems like postgres-operator (or spilo for all I know) does not populate that file even though the var is present in the configmap...

If i do this:
```
 echo -ne "eu-west-1" > /run/etc/wal-e.d/env/AWS_REGION
```

Then archiving does work:
```
postgres@team-cluster-0:~/pgdata/pgroot/data/pg_wal$ ls archive_status/
000000010000000000000001.done                  000000010000000000000002.done                  000000010000000000000003.done  000000010000000000000005.00000028.backup.done  000000010000000000000006.done
000000010000000000000002.000000D0.backup.done  000000010000000000000003.00000028.backup.done  000000010000000000000004.done  000000010000000000000005.done
postgres@team-cluster-0:~/pgdata/pgroot/data/pg_wal$ 
```

I could manually create a configmap to populate that file manually however, i would prefer understanding how postgres-operator (/spilo?) could be able to create that file by itself.

Thanks a lot for any help

**Comment by vfrans** [comment]
Apparently for some reasons **not** specifying does fix the issue and everything is working as expected once it is removed...

**Comment by FxKu** [comment]
I hope it's proper documented now with #1592 merged.

**Comment by ienliven** [comment]
> envdir "/run/etc/wal-e.d/env" wal-g backup-list

**Comment by estein9825** [comment]
This might help those who came by here. I ran into the same issue with having an assumed role in my service account that should have made it worked. It turned out, however, that spilo uses an instance profile to get the account info and not the sts assumed role. See https://github.com/zalando/spilo/blob/4731ce3f26ace36d6c7687a194f4055edce2dc7e/postgres-appliance/scripts/wal-e-wal-fetch.sh#L8.

As a result, the account that was used when a backup was run manually was different from the one used by spilo to fetch the wal fragments in S3. Once I granted S3 access to that instance profile, the archive/restore commands were happy.

At the time of this posting, the only options was AWS Key/Secret or Instance Proifle. There was no assume role ability.

I opened up an issue related to it: https://github.com/zalando/spilo/issues/837
